components/fxa-client/src/internal/state_manager.rs

```diff
@@ -0,0 +1,213 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::{HashMap, HashSet};
+
+use crate::{
+    internal::{
+        commands, device,
+        oauth::{AccessTokenInfo, RefreshToken},
+        profile::Profile,
+        state_persistence::state_to_json,
+        CachedResponse, Config, OAuthFlow, PersistedState,
+    },
+    Result, ScopedKey,
+};
+
+/// Stores and manages the current state of the FxA client
+///
+/// All fields are private, which means that all state mutations must go through this module.  This
+/// makes it easier to reason about state changes.
```

Comment:
I absolutely love hiding the fields and only exposing the functions 🙏 This is awesome!